### PR TITLE
MATX-236 - README.md: reloading libraries, configurable resource search paths and intermediate file dump path.

### PR DESCRIPTION
**Pull request by ppenenko**
…d intermediate file dump path.

### CODE REVIEW

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 1]
```diff
@@ -2,11 +2,13 @@
...
 
-This module contains code for a sample Maya plug-in which implements Maya surface and texture shading nodes representing MaterialX shaders in Maya's OpenGL viewport (VP2.0). To support this functionality, MaterialX elements are compiled into GLSL code as well as XML shade fragment wrappers required by Maya to integrate the fragments into the final viewport shader.
+This module contains code for a sample Maya plug-in which implements Maya surface and texture shading nodes representing MaterialX nodes in Maya's OpenGL viewport (VP2.0).
+To support this functionality, MaterialX elements are compiled into GLSL code as well as XML shade fragment wrappers required by Maya to integrate the fragments into the final viewport shader.
 
 ## Build
 
-To build this plug-in, the `MATERIALX_BUILD_CONTRIB` build flag must be enabled and the respective `MAYA_DEBUG_DIR` and/or `MAYA_RELEASE_DIR` build directory options set. The minimum required Maya version is Maya 2019.
+To build this plug-in, the `MATERIALX_BUILD_CONTRIB` build flag must be enabled and the respective `MAYA_DEBUG_DIR` and/or `MAYA_RELEASE_DIR` build directory options must be set.
+The minimum required Maya version is Maya 2019.
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: Such line breaks make no difference in the rendered Markdown but make lines shorter and diffs clearer in some tools.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 2]
```diff
@@ -45,44 +49,48 @@ Modifying any one of these attributes via GUI or scripting causes the shading no
 ## Maya Commands
 ### `createMaterialXNode`
 
-Creates one or many MaterialX nodes from the specified MaterialX document. The command accepts the following flags:
+Creates one or many MaterialX nodes from the specified MaterialX document.
+The command accepts the following flags:
 
-```MEL
+```Text
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: Not technically a MEL snippet - it's copied from what `help createMaterialXNode` outputs in the log.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 3]
```diff
@@ -92,21 +100,49 @@ For example:
 reloadMaterialXNode Jade_SR_jade
 ```
 
-Reloads the document and refreshes the viewport shader mapped to the specified MaterialX node. This command is useful when the contents of the document file have changed on disk, e.g. due to editing in an external XML editor.
+Reloads the document and refreshes the viewport shader mapped to the specified MaterialX node.
+This command is useful when the contents of the document file have changed on disk, e.g. due to editing in an external XML editor.
+
+### `reloadMaterialXLibraries`
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: Documented a new command.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 4]
```diff
@@ -92,21 +100,49 @@ For example:
...
+
+This command has no arguments.
+When executed, it reloads all MaterialX libraries used by the plug-in (see below for details).
 
 ## MaterialX Libraries
 
-The plug-in loads a set of standard MaterialX libraries (installed with the plug-in) at load time. When a document is loaded for a MaterialX node, it imports these libraries from memory instead of reading them from the file system.
+MaterialX documents can reference elements defined in other documents called libraries.
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: An introduction into libraries to make the reloading comments below clearer.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 5]
```diff
@@ -92,21 +100,49 @@ For example:
...
+Here's an example of how the two option variables can be configured in MEL:
 
 ```MEL
-optionVar -stringValueAppend materialXLibrarySearchPaths "D:/MaterialX/Teapot_Demo_MatX"
+optionVar -stringValueAppend materialXLibrarySearchPaths "D:/MaterialX/Demo/Libraries"
 optionVar -stringValueAppend materialXLibraryNames "stdlib"
 ```
 
-## Editor integration
+Please note that library and per-node document loading are independent of each other. When a MaterialX document is loaded for a `MaterialXSurfaceNode` or `MaterialXTextureNode`, it imports elements from the libraries currently in memory which speeds up document loading.
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: About reloading.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 6]
```diff
@@ -92,21 +100,49 @@ For example:
...
+
+On the other hand, all libraries can be reloaded explicitly with the `reloadMaterialXLibraries` Maya command which does not affect the currently loaded documents. The updated libraries will be imported by documents loaded or reloaded afterwards.
+
+## Resource Search Paths
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: Another new feature: resource search paths.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 7]
```diff
@@ -116,6 +152,20 @@ optionVar -stringValue materialXEditorPath "D:/MyEditor/bin/MyEditor.exe"
 
 To open the MaterialX document in the editor, press **Edit** in the Attribute Editor of the MaterialX node. Once the document has been edited and saved, press **Reload** in the Attribute Editor which executes the `reloadMaterialXNode` command and refreshes the shader in the viewport.
 
+## Debugging
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: Another new section.

Comment (y2yeung) [reply source/MaterialXContrib/MaterialXMaya/README.md]: May I suggest a slight rewording of the 2nd sentence:
"The Maya API allows plug-ins to configure paths to which these intermediate files can be saved for debugging purposes."
(Are the two kinds of intermediate files: 1) the complex shading graph and 2) the final GLSL shading effect?)

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 5 above)
Comment (y2yeung) [reply source/MaterialXContrib/MaterialXMaya/README.md]: No need to be so polite and say 'please' in the docs. :) 
How about:
The loading of libraries and the loading of documents for a MaterialXSurfaceNode or MaterialXTextureNode are independent of each other. In the latter case, the plug-in imports elements from the libraries currently in memory, which speeds up document loading.

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 6 above)
Comment (y2yeung) [reply source/MaterialXContrib/MaterialXMaya/README.md]: How about "The updated libraries for these documents will be imported when the documents themselves are loaded or reloaded."

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 5 above)
Comment (ppenenko) [reply source/MaterialXContrib/MaterialXMaya/README.md]: Agreed, I'll make the change.

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 7 above)
Comment (ppenenko) [reply source/MaterialXContrib/MaterialXMaya/README.md]: Makes sense.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 8]
```diff
@@ -116,6 +152,20 @@ optionVar -stringValue materialXEditorPath "D:/MyEditor/bin/MyEditor.exe"
...
+
+When a Maya plug-in shading node provides a shade fragment to Maya, it's composited with Maya's standard fragments to form a complex shading graph, which in its turn is used to generate the final GLSL shader effect to render with in the viewport.
+Maya API [allows](https://help.autodesk.com/view/MAYAUL/2019/ENU/?guid=Maya_SDK_MERGED_cpp_ref_class_m_h_w_render_1_1_m_fragment_manager_html) plug-ins to configure paths where to save these intermediate files for debugging purposes.
```
Comment (bernardkwok) [inline source/MaterialXContrib/MaterialXMaya/README.md]: Minor: The Maya API...

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 2 above)
Comment (bernardkwok) [reply source/MaterialXContrib/MaterialXMaya/README.md]: Does ```MEL even mean anything in Markdown?

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 6 above)
Comment (ppenenko) [reply source/MaterialXContrib/MaterialXMaya/README.md]: I think "afterwards" is the important part here (I would welcome a better expression though): once the libraries have been updated, they are going to be used by all subsequent document "loads". Also, the documents in this sentence are not necessarily the same as in the "currently loaded documents" in the previous sentence.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 9]
```diff
@@ -2,13 +2,15 @@
 
 ## Description
 
-This module contains code for a sample Maya plug-in which implements Maya surface and texture shading nodes representing MaterialX shaders in Maya's OpenGL viewport (VP2.0). To support this functionality, MaterialX elements are compiled into GLSL code as well as XML shade fragment wrappers required by Maya to integrate the fragments into the final viewport shader.
+This module contains code for a sample Maya plug-in which implements Maya surface and texture shading nodes representing MaterialX nodes in Maya's OpenGL viewport (VP2.0).
+To support this functionality, MaterialX elements are compiled into GLSL code as well as XML shade fragment wrappers required by Maya to integrate the fragments into the final viewport shader.
 
-## Build
+## Building
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: An unrelated minor fix.

File: source/MaterialXContrib/MaterialXMaya/README.md [hunk 10]
```diff
@@ -2,13 +2,15 @@
...
 
-To build this plug-in, the `MATERIALX_BUILD_CONTRIB` build flag must be enabled and the respective `MAYA_DEBUG_DIR` and/or `MAYA_RELEASE_DIR` build directory options set. The minimum required Maya version is Maya 2019.
+To build this plug-in, the `MATERIALX_BUILD_CONTRIB` build flag must be enabled and the respective `MAYA_DEBUG_DIR` and/or `MAYA_RELEASE_DIR` build directory options must be set.
+The minimum required Maya version is Maya 2019.
 
-## Load
+## Loading
```
Comment (ppenenko) [inline source/MaterialXContrib/MaterialXMaya/README.md]: An unrelated minor fix.

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 8 above)
Comment (ppenenko) [reply source/MaterialXContrib/MaterialXMaya/README.md]: Right, corrected.

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 2 above)
Comment (ppenenko) [reply source/MaterialXContrib/MaterialXMaya/README.md]: Yes. This can be the name of any programming language though not all renderers support all languages. Visual Studio Code knows MEL syntax via the MayaCode extension and so it's able to apply MEL syntax highlighting in Markdown preview as well. GitHub sadly doesn't know about MEL but we could fix this by contributing to https://github.com/github/linguist/blob/master/lib/linguist/languages.yml

File: source/MaterialXContrib/MaterialXMaya/README.md (hunk 6 above)
Comment (y2yeung) [reply source/MaterialXContrib/MaterialXMaya/README.md]: Why don't we just use the words that you suggested "subsequent document loads": 

The updated libraries will be imported upon subsequent document loading and reloading.

Comment (ppenenko) [reply source/MaterialXContrib/MaterialXMaya/README.md]: What do you think about the fix I just committed? Just changing "afterwards" to "subsequently". I think "imported by documents" is more specific because libraries are imported to or by documents.

Comment (y2yeung) [reply source/MaterialXContrib/MaterialXMaya/README.md]: How about: "The updated libraries will be imported by documents that are subsequently loaded or reloaded." ?
